util/include/crypto/openssl/integer.hpp

```diff
@@ -24,121 +24,121 @@
   Integer operator*(const Integer& i) const {
-    BN_CTX* ctx = BN_CTX_new();
-    if (!ctx) THROW_OPENSSL_ERROR;
-    concord::util::ScopeExit s{[&]() { BN_CTX_free(ctx); }};
-    BIGNUM* res = BN_new();
-    if (!res) THROW_OPENSSL_ERROR;
-    if (!BN_mul(res, num_, i.num_, ctx)) THROW_OPENSSL_ERROR;
-    return res;
+    UniqueBNCTX ctx(BN_CTX_new());
+    assertResultValid(ctx.get());
+    Integer result;
+    assertResultValid(BN_mul(result.num_.get(), num_.get(), i.num_.get(), ctx.get()));
+    return result;
   }
   Integer operator/(const Integer& i) const {
-    BN_CTX* ctx = BN_CTX_new();
-    if (!ctx) THROW_OPENSSL_ERROR;
-    concord::util::ScopeExit s{[&]() { BN_CTX_free(ctx); }};
-    BIGNUM* res = BN_new();
-    if (!res) THROW_OPENSSL_ERROR;
-    if (!BN_div(res, nullptr, num_, i.num_, ctx)) THROW_OPENSSL_ERROR;
-    return res;
+    UniqueBNCTX ctx(BN_CTX_new());
+    assertResultValid(ctx.get());
+    Integer result;
+    assertResultValid(BN_div(result.num_.get(), nullptr, num_.get(), i.num_.get(), ctx.get()));
+    return result;
   }
   Integer operator%(const Integer& i) const {
-    BN_CTX* ctx = BN_CTX_new();
-    if (!ctx) THROW_OPENSSL_ERROR;
-    concord::util::ScopeExit s{[&]() { BN_CTX_free(ctx); }};
-    BIGNUM* res = BN_new();
-    if (!res) THROW_OPENSSL_ERROR;
-    if (!BN_mod(res, num_, i.num_, ctx)) THROW_OPENSSL_ERROR;
-    return res;
+    UniqueBNCTX ctx(BN_CTX_new());
+    assertResultValid(ctx.get());
+    Integer result;
+    assertResultValid(BN_mod(result.num_.get(), num_.get(), i.num_.get(), ctx.get()));
+    return result;
   }
   Integer operator=(const Integer& i) {
-    num_ = BN_dup(i.num_);
-    if (!num_) THROW_OPENSSL_ERROR;
+    num_.reset(BN_dup(i.num_.get()));
     return *this;
   }
 
-  bool operator!=(const Integer& i) const { return BN_cmp(num_, i.num_) != 0; }
-  bool operator==(const Integer& i) const { return BN_cmp(num_, i.num_) == 0; }
+  bool operator!=(const Integer& i) const { return BN_cmp(num_.get(), i.num_.get()) != 0; }
+  bool operator==(const Integer& i) const { return BN_cmp(num_.get(), i.num_.get()) == 0; }
 
   Integer operator<<(size_t n) const {
-    BIGNUM* res = BN_new();
-    if (!res) THROW_OPENSSL_ERROR;
-    if (!BN_lshift(res, num_, n)) THROW_OPENSSL_ERROR;
-    return res;
+    Integer result;
+    if (!BN_lshift(result.num_.get(), num_.get(), n)) THROW_OPENSSL_ERROR;
+    return result;
   }
-  size_t size() const { return BN_num_bytes(num_); }
+
+  size_t size() const { return BN_num_bytes(num_.get()); }
   std::string toHexString(bool add_prefix = false) const {
-    if (add_prefix) return std::string("0x") + BN_bn2hex(num_);
-    return BN_bn2hex(num_);
+    std::string result{UniqueOpenSSLString{BN_bn2hex(num_.get())}.get()};
+    if (add_prefix) {
+      result = std::string("0x") + result;
+    }
+    return result;
   }
-  std::string toDecString() const { return BN_bn2dec(num_); }
+  std::string toDecString() const { return BN_bn2dec(num_.get()); }
```

Comment:
Shouldn't it use UniqueOpenSSLString as well?

Reply:
Thanks, I'll fix it on the other PR to master (https://github.com/vmware/concord-bft/pull/2861). Then when it's ready will create new PR to [v0.16.x](https://github.com/vmware/concord-bft/tree/v0.16.x)